tool/openapi/openapi_spec.go

```diff
@@ -0,0 +1,120 @@
+//
+// Tencent is pleased to support the open source community by making trpc-agent-go available.
+//
+// Copyright (C) 2025 Tencent.  All rights reserved.
+//
+// trpc-agent-go is licensed under the Apache License Version 2.0.
+//
+//
+
+package openapi
+
+import (
+	"context"
+	"io"
+	"net/url"
+
+	openapi "github.com/getkin/kin-openapi/openapi3"
+)
+
+// SpecLoader loads the OpenAPI spec.
+type SpecLoader interface {
+	// Load loads the OpenAPI spec.
+	Load(ctx context.Context) (*openapi.T, error)
+}
+
+type dataSpecLoader struct {
+	r io.Reader
+}
+
+// Load loads the OpenAPI spec from io.Reader.
+func (d *dataSpecLoader) Load(ctx context.Context) (*openapi.T, error) {
+	loader := openapi.Loader{Context: ctx}
+	return loader.LoadFromIoReader(d.r)
+}
+
+// NewDataLoader creates a new data spec loader.
+func NewDataLoader(r io.Reader) *dataSpecLoader {
+	return &dataSpecLoader{r: r}
+}
+
+type fileSpecLoader struct {
+	path string
+}
+
+// Load loads the OpenAPI spec from file.
+func (f *fileSpecLoader) Load(ctx context.Context) (*openapi.T, error) {
+	loader := openapi.Loader{Context: ctx}
+	return loader.LoadFromFile(f.path)
+}
+
+// NewFileLoader creates a new file spec loader.
+func NewFileLoader(filePath string) *fileSpecLoader {
+	return &fileSpecLoader{path: filePath}
+}
+
+type urlSpecLoader struct {
+	uri string
+}
+
+// Load loads the OpenAPI spec from url.
+func (u *urlSpecLoader) Load(ctx context.Context) (*openapi.T, error) {
+	uri, err := url.Parse(u.uri)
+	if err != nil {
+		return nil, err
+	}
+	loader := openapi.Loader{Context: ctx}
+	return loader.LoadFromURI(uri)
+}
+
+// NewURILoader creates a new url spec loader.
+func NewURILoader(uri string) *urlSpecLoader {
+	return &urlSpecLoader{uri: uri}
+}
+
+func newSpecProcessor(doc *openapi.T) *specProcessor {
+	return &specProcessor{doc: doc}
+}
+
+type specProcessor struct {
+	doc        *openapi.T
+	operations []*Operation
+}
+
+// InfoTitle returns the title of the spec.
+func (s *specProcessor) InfoTitle() string {
+	return s.doc.Info.Title
+}
+
+const defaultBaseURL = "/"
+
+func (s *specProcessor) processOperations() error {
+	baseURL := defaultBaseURL
+	if len(s.doc.Servers) != 0 {
+		baseURL = s.doc.Servers[0].URL
```

Comment:
2. Logic: Base URL Handling
       * Issue: The code defaults to s.doc.Servers[0].URL if servers exist. OpenAPI Server URLs often contain variables
         (e.g., https://{username}.example.com:{port}/api). Using the raw URL string might not be what the user expects.
       * Suggestion: Consider documenting this limitation or implementing variable expansion if supported by the
         underlying library or requirements. Additionally, silently picking the first server is acceptable but might
         warrant a comment or log if multiple servers are defined.

Reply:
https://learn.openapis.org/specification/servers.html#server-variables